### ISSUE TITLE
Alerting: Incompatibilities with database - AWS RDS, `aurora_mysql` backend

### ISSUE THREAD

**Issue by alexweav** [grafana/grafana#54556]
**What happened**:

A user observed some 500s when using Grafana against AWS RDS, on the `aurora_mysql` backend.

After enabling database logging, the SQL statement that emitted the failure was logged as:
```
INSERT INTO alert_configuration
(alertmanager_configuration, configuration_hash, configuration_version, org_id, created_at, %s) 
SELECT T.* FROM (SELECT ? AS alertmanager_configuration,? AS configuration_hash,? AS configuration_version,? AS org_id,? AS created_at,? AS 'default') AS T
WHERE
EXISTS (
	SELECT 1 
	FROM alert_configuration 
	WHERE 
		org_id = ? 
	AND 
		id = (SELECT MAX(id) FROM alert_configuration WHERE org_id = ?) 
	AND 
		configuration_hash = ?
)
```
taken from `store/alertmanager.go` in unified alerting.

This failed with the error:
```
Error 1051: Unknown table 'T'
```

It seems that subquery aliases do not work on RDS's `aurora_mysql` backend. This same statement works correctly on MySQL natively, and on GCP Cloud SQL's MySQL interface. This appears to be an incompatibility between the three.

**What you expected to happen**:

I expected the request to not return a 500.

**How to reproduce it (as minimally and precisely as possible)**:

1. Use AWS RDS with `aurora_mysql` backend as the Grafana database.
2. Enable unified alerting.
3. Hit any endpoint that inserts Alertmanager configurations. In my case I used the provisioning API for contact points.

**Anything else we need to know?**:

**Environment**:
- Grafana version: 9.1.x, main
- Data source type & version: n/a
- OS Grafana is installed on: n/a
- User OS & Browser: n/a
- Grafana plugins: n/a
- Others: Database in MySQL mode, pointing to AWS RDS with `aurora_mysql` backend.

**Comment by joeblubaugh** [comment]
The Grafana-supported databases are listed here: https://grafana.com/docs/grafana/latest/setup-grafana/installation/#supported-databases

RDS/MySQL isn't officially supported, but we can look into whether re-writing this query would be enough to avoid the issue. We should check for any subquery aliases we can find, while we're at it.

**Comment by wiardvanrij** [comment]
AWS claims (and to be really honest, I believe them, unless we are suddenly the only ones doing this?) that they are 100% compatible with MySQL. So, I'm actually going after that and figure out if it is even an AWS RDS issue. That said, when I saw the query, I was questioning if that is the nicest query we could do.

**Comment by wiardvanrij** [comment]
Ok, I did some testing. It only gave me more questions, but perhaps it helps. 

```
mysql> SELECT T.* FROM (SELECT configuration_hash as foo FROM alert_configuration as T);
ERROR 1248 (42000): Every derived table must have its own alias
mysql>  SELECT T.* FROM (SELECT configuration_hash as foo FROM alert_configuration as T) as B;
ERROR 1051 (42S02): Unknown table 'T'
mysql> SELECT B.* FROM (SELECT configuration_hash as foo FROM alert_configuration as T) as B;
+----------------------------------+
| foo                              |
+----------------------------------+
| 8c409350c88d78d2ee938448449e628d |
| 5d9521fbbc7d80444f406fd0273dfa43 |
| 0d55db0e1fa2e8da91af04ecf7523b7e |
| 1c2815c03601c302f72854fc0572d9d1 |
+----------------------------------+
4 rows in set (0.00 sec)
```

Then I migrated the DB from AWS RDS into a native MySQL database using the latest image from mysql:

```
mysql> SELECT T.* FROM (SELECT configuration_hash as foo FROM alert_configuration as T);
ERROR 1248 (42000): Every derived table must have its own alias
mysql>  SELECT T.* FROM (SELECT configuration_hash as foo FROM alert_configuration as T) as B;
ERROR 1051 (42S02): Unknown table 'T'
mysql> SELECT B.* FROM (SELECT configuration_hash as foo FROM alert_configuration as T) as B;
+----------------------------------+
| foo                              |
+----------------------------------+
| 8c409350c88d78d2ee938448449e628d |
| 5d9521fbbc7d80444f406fd0273dfa43 |
| 0d55db0e1fa2e8da91af04ecf7523b7e |
| 1c2815c03601c302f72854fc0572d9d1 |
+----------------------------------+
4 rows in set (0.00 sec)
```

Exactly the same result.


So then I would think if the query is altered into:

```
INSERT INTO alert_configuration
(alertmanager_configuration, configuration_hash, configuration_version, org_id, created_at, %s) 
SELECT B.* FROM (SELECT ? AS alertmanager_configuration,? AS configuration_hash,? AS configuration_version,? AS org_id,? AS created_at,? AS 'default') AS T
WHERE
EXISTS (
	SELECT 1 
	FROM alert_configuration 
	WHERE 
		org_id = ? 
	AND 
		id = (SELECT MAX(id) FROM alert_configuration WHERE org_id = ?) 
	AND 
		configuration_hash = ?
) as B
```

It should work? (note the replace from `T.*` to `B.*` and the addition of `as B` on the end. 

Yet I'm more curious on why nobody else could reproduce the behaviour on your MySQL systems as it is somewhat clearly for just myself, that it cannot work? I.e. the result I'm getting makes more sense. 

Unless I made a mistake in my query somewhere?

**Comment by joeblubaugh** [comment]
What version of "native MySQL" are you testing with? We should be sure to test against the same version.

On MySQL 5.7.39:
```
> SELECT T.* FROM (SELECT configuration_hash as foo FROM alert_configuration as T);
Query 1 ERROR: Every derived table must have its own alias

> SELECT T.* FROM (SELECT configuration_hash as foo FROM alert_configuration) as T;
(Returns a result)
```

The second query succeeds, and to my eyes matches the sub-query in the Grafana source code, so I don't think this is the root of the RDS issue.

**Comment by wiardvanrij** [comment]
`mysql  Ver 8.0.30 for Linux on x86_64 (MySQL Community Server - GPL)` 

and yes, your second query works there as well:
```
mysql> SELECT T.* FROM (SELECT configuration_hash as foo FROM alert_configuration) as T;
(Returns a result)
```

It also works on RDS

**Comment by wiardvanrij** [comment]
So my query was bad. Yet I still don't entirely understand this part of the query:

```
SELECT T.* FROM (SELECT ? AS alertmanager_configuration,? AS configuration_hash,? AS configuration_version,? AS org_id,? AS created_at,? AS 'default') AS T
```

How are we doing a select in a select, without defining where the select is FROM? 

The query gets followed up by a:
```
WHERE
EXISTS (
	SELECT 1 
	FROM alert_configuration 
	WHERE 
		org_id = ? 
	AND 
		id = (SELECT MAX(id) FROM alert_configuration WHERE org_id = ?) 
	AND 
		configuration_hash = ?
)
```

If you could craft me the exact query that is being made, then I also can test it against RDS.

**Comment by joeblubaugh** [comment]
I don't understand what you want us to generate. You can make this a "raw" query by replacing each ? with values of the appropriate type for the columns. To try the entire query you'll need to add at least one row first with a simple INSERT statement.

That's because this complex INSERT adds a row to the alert_configuration table when a new configuration is created. The caller needs to have the configuration hash of the latest version. That's what the `WHERE EXISTS` clause is checking: Has the latest configuration hash changed since the caller last read it? If so, abandon the configuration write.

-----

```
(SELECT ? AS alertmanager_configuration,? AS configuration_hash,? AS configuration_version,
  ? AS org_id,? AS created_at,? AS 'default') AS T
```
is creating a derived table T with the column values bound at query time into the ?. I'm not sure this is actually necessary, since
```
INSERT INTO alert_configuration
(alertmanager_configuration, configuration_hash, configuration_version, org_id, created_at, %s) 
VALUES 
(?, ?, ?,? ,? ,? )
```
seems equivalent, but I need to investigate that further.

**Comment by wiardvanrij** [comment]
> I don't understand what you want us to generate

Two reasons;

1: I'm terrible at SQL and my previous attempt to reproduce it with 'native MySQL' and Aurora has proven that I did the wrong query in the first place
2: To have something 'outside' Grafana that can reproduce it working on 'native MySQL' but failing on Aurora. This makes it easier to pin-point _what_ actually fails in Aurora.

**Comment by wiardvanrij** [comment]
For the record, I've migrated from Aurora to AWS RDS and it's working fine now.

**Comment by JacobsonMT** [comment]
Might be a good idea to update the [docs](https://grafana.com/docs/grafana/latest/setup-grafana/installation/#supported-databases) to mention that Aurura with MySQL compatibility is not supported. @brendamuir

**Comment by armandgrillet** [comment]
There are already some details in the docs you linked with the sentence "Grafana can report errors when relying on read-only MySQL servers, such as in high-availability failover scenarios or serverless AWS Aurora MySQL. This is a known issue; for more information, see [issue #13399](https://github.com/grafana/grafana/issues/13399).", @brendamuir shall we rephrase this to mention this issue as well?

**Comment by brendamuir** [comment]
I updated the docs here: https://github.com/grafana/grafana/pull/56630/files.

**Comment by mgeggie** [comment]
Hello,

We manage a number of our alerts and dashboards on our Grafana deployments via the Grafana Terraform provider, and through our upgrade of that Terraform codebase to support the new features introduced with Grafana alertmanager we also discovered that our Grafana deployments of 5+ years on AWS Aurora MySQL now need to be migrated. It seems odd to me that managing alertmanager features via the Grafana UI does work for us on Aurora MySQL, but that the API endpoints (used by Terraform) to manage individual alertmanager objects are the only ones failing.

In any case, due to this change, we've now been forced into migrating our Grafana database of 5+ years from AWS Aurora MySQL to vanilla AWS RDS MySQL to support the new alertmanager feature, which will be made mandatory as-of the upcoming Grafana 10.

Simply updating the documentation that alertmanager alone doesn't support AWS RDS Aurora MySQL seems like a fading fine line to walk without explicitly dropping support for Aurora MySQL in Grafana at-large when the feature causing that incompatibility is entering the mainline product featureset in the next major release.

I'd ask that the team reconsider investigating this issue with a more focused attention towards resolving the compatibility with Aurora MySQL, rather than dropping support for a database platform that's been supported by the product going back over a half decade.

Thanks for your (hopeful) reconsideration of this issue.

**Comment by Bak3y** [comment]
Bump and second @mgeggie's comment - please reopen this issue.  Not supporting AWS RDS for a core component of Grafana isn't really an option.

**Comment by alexweav** [comment]
We're looking into this, but note for the time being that we _do_ support AWS RDS MySQL. This is different from AWS RDS Aurora DB with MySQL Compatibility layer.

You can work around this by choosing native MySQL in AWS, rather than Aurora with MySQL Compatibility. It seems the advertised "MySQL Compatibility" is not 100% compatible, which is unfortunate.

**Comment by alexweav** [comment]
Oh! And, I just realized, the [database schema change](https://github.com/grafana/grafana/issues/60053) that I'm actively working on is **very** likely to automatically solve this. It completely changes the SQL here that's giving us issues here, to something much simpler that the compatibility layer is more likely to not fail on. Let's test again when that rework is complete, because it seems very likely that we'll get this for free.

**Comment by alexweav** [comment]
Although one change got rolled back, the last PR for the above is now up.

I set up an aurora DB with mysql compat, on main with this PR added: https://github.com/grafana/grafana/pull/60492

And, I can confirm that with that change, Unified Alerting now works correctly with RDS Aurora!
I'm re-opening this issue, so we remember to remove that statement from the docs.

**Comment by wiardvanrij** [comment]
> Although one change got rolled back, the last PR for the above is now up.
> 
> I set up an aurora DB with mysql compat, on main with this PR added: #60492
> 
> And, I can confirm that with that change, Unified Alerting now works correctly with RDS Aurora! I'm re-opening this issue, so we remember to remove that statement from the docs.

Thanks for your work. My only "concern" or question basically is, will Grafana keep this in mind for future updates. I.e. will there be e2e tests or actively considered that each new feature/change must/should work agains Aurora? 
This to maintain its support and consistency in the future?
